Sources/SwiftProtobuf/Message+AsyncMessageSequence.swift

```diff
@@ -0,0 +1,232 @@
+  @frozen public struct AsyncIterator: AsyncIteratorProtocol, Sendable
+  where Base.AsyncIterator: Sendable {
+    
+    private var iterator: Base.AsyncIterator?
+    private let extensions: ExtensionMap?
+    private let partial: Bool
+    private let options: BinaryDecodingOptions
+    
+    init(
+      iterator: Base.AsyncIterator,
+      extensions: ExtensionMap?,
+      partial: Bool,
+      options: BinaryDecodingOptions
+    ) {
+      self.iterator = iterator
+      self.extensions = extensions
+      self.partial = partial
+      self.options = options
+    }
+    
+    /// Aysnchronously reads the next varint
+    private mutating func nextVarInt() async throws -> UInt64? {
+      var messageSize: UInt64 = 0
+      var shift: UInt64 = 0
+      
+      while let byte = try await iterator?.next() {
+        messageSize |= UInt64(byte & 0x7f) << shift
+        shift += UInt64(7)
+        if shift > 35 {
+          throw BinaryDecodingError.malformedProtobuf
+        }
+        if (byte & 0x80 == 0) {
+          return messageSize
+        }
+      }
+      if (shift > 0) {
+        // The stream has ended inside a varint.
+        throw BinaryDecodingError.truncated
+      }
+      return nil // End of stream reached.
+    }
+    
+    /// Asynchronously advances to the next message and returns it, or ends the
+    /// sequence if there is no next message.
+    ///
+    /// - Returns: The next message, if it exists, or `nil` to signal the end of
+    ///   the sequence.
+    public mutating func next() async throws -> M? {
+      guard let messageSize = try await nextVarInt() else {
+        iterator = nil
+        return nil
+      }
+      if messageSize == 0 {
+        return try M(
+          serializedBytes: [],
+          extensions: extensions,
+          partial: partial,
+          options: options
+        )
+      } else if messageSize > 0x7fffffff {
+        throw BinaryDecodingError.tooLarge
+      }
+      
+      var buffer = [UInt8](repeating: 0, count: Int(messageSize))
+      var consumedBytes = 0
+      
+      while let byte = try await iterator?.next() {
+        buffer[consumedBytes] = byte
+        consumedBytes += 1
+        if consumedBytes == messageSize {
+          return try M(
+            serializedBytes: buffer,
+            extensions: extensions,
+            partial: partial,
+            options: options
+          )
+        }
+      }
+      throw BinaryDecodingError.truncated // The buffer was not filled.
+    }
+  }
+  
+  /// Creates the asynchronous iterator that produces elements of this
+  /// asynchronous sequence.
+  ///
+  /// - Returns: An instance of the `AsyncIterator` type used to produce
+  /// messages in the asynchronous sequence.
+  public func makeAsyncIterator() -> AsyncMessageSequence.AsyncIterator {
+    AsyncIterator(
+      iterator: baseSequence.makeAsyncIterator(),
+      extensions: extensions,
+      partial: partial,
+      options: options
+    )
+  }
+}
```

Comment:
Can we add an explicit `Sendable` unavailability to the iterator here please

```
@available(*, unavailable)
extension AsyncMessageSequence.Iterator: Sendable { }
```

Reply:
Done.